klib/cloud_init.c

```diff
@@ -269,21 +271,21 @@ closure_function(7, 1, status, cloud_download_recv,
             deallocate_closure(bound(vh));
         closure_finish();
     }
-    return STATUS_OK;
+    return false;
   error:
     apply(bound(out), 0);   /* close connection */
-    return STATUS_OK;
+    return true;
 }
 
 static boolean cloud_download_retry(connection_handler ch);
 
-closure_function(1, 1, buffer_handler, cloud_download_ch,
+closure_function(1, 1, input_buffer_handler, cloud_download_ch,
                  cloud_download_cfg, cfg,
                  buffer_handler, out)
 {
     cloud_download_cfg cfg = bound(cfg);
     status_handler sh = (status_handler)&cfg->complete;
-    buffer_handler in = 0;
+    input_buffer_handler in = 0;
```

Comment:
Should this actually be initialized to INVALID_ADDRESS as in cloud_azure.c? If there is a connection established but some other condition (in this particular case I think it would be transmit failure on output buffer handler) causes it to skip initialization of 'in', which would return 0, which is not checked in direct_receive_service, and perhaps could lead to trying to apply a null pointer if there were something in the receive queue. Maybe something about this particular case could prevent that from happening, but seems like INVALID_ADDRESS is the safer initialized value? Also, same question in a one other place.

Reply:
Yes, I agree, INVALID_ADDRESS is a safer initialization value. I fixed that, both here and in klib/radar.c